File: AI-LAB/AgentGold.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

using namespace std;

bool isValid(int x, int y, int N) {
    return 0 <= x && x < N && 0 <= y && y < N;
}

void agentProgram(int N, int X, pair<int, int> chargingStation, pair<int, int>& potOfGold) {
    int x = chargingStation.first;
    int y = chargingStation.second;
    int timeTaken = 0;
    int charge = X;

    vector<vector<bool>> explored(N, vector<bool>(N, false));

    while (charge > 0) {
        //  Simulating the current location of an agent
        cout << "Current location: (" << x << ", " << y << ")" << endl;

        // Checking if the agent found the gold
        if (make_pair(x, y) == potOfGold) {
            cout << "Pot of gold found!" << endl;
            break;
        }

        // If room explored making it true
        explored[x][y] = true;

        // Move the agent based on the rules
        if (isValid(x + 1, y, N) && !explored[x + 1][y]) {
            x = x + 1;  // Move forward
        } else if (isValid(x, y + 1, N) && !explored[x][y + 1]) {
            y = y + 1;  // Move right
        } else if (isValid(x - 1, y, N) && !explored[x - 1][y]) {
            x = x - 1;  // Move backward
        } else if (isValid(x, y - 1, N) && !explored[x][y - 1]) {
            y = y - 1;  // Move left
        } else {
            // If all neighboring rooms are explored, move backward
            x = x - 1;
        }

        // Update charge and time taken
        charge--;
        timeTaken++;
    }

    // Output total time taken
    // Here total time I assumed to be the no. of steps or moves the agent makes 
    // while navigating the grid to find the pot of gold or until it decides to go to the charging station.
    cout << "Total time taken: " << timeTaken << " units" << endl;

    // Output the location where the decision to charge was taken
    cout << "Decision to charge taken at: (" << x << ", " << y << ")" << endl;
}

int main() {
    srand(time(0));  

    // Inputs
    int N, X;
    cout << "Enter the size of the grid (N): ";
    cin >> N;
    cout << "Enter the battery capacity (X): ";
    cin >> X;

    // Charging station location input
    pair<int, int> chargingStation;
    cout << "Enter the charging station location (x y): ";
    cin >> chargingStation.first >> chargingStation.second;

    // Pot of gold location input
    pair<int, int> potOfGold;
    cout << "Enter the pot of gold location (x y): ";
    cin >> potOfGold.first >> potOfGold.second;

    // Run the agent program
    agentProgram(N, X, chargingStation, potOfGold);

    return 0;
}

File: AI-LAB/AgentGold.py


import random

def is_valid(x, y, N):
    return 0 <= x < N and 0 <= y < N

def agent_program(N, X, charging_station, pot_of_gold):
    x, y = charging_station
    time_taken = 0
    charge = X

    explored = [[False for _ in range(N)] for _ in range(N)]

    while charge > 0:
        # Simulating the current location of an agent
        print(f"Current location: ({x}, {y})")

        # Checking whether Gold Found by an agent
        if (x, y) == pot_of_gold:
            print("GOLD found !!")
            break

        # If room explored making it true
        explored[x][y] = True

        # Move the agent based on the rules
        if is_valid(x + 1, y, N) and not explored[x + 1][y]:
            x = x + 1  # Move forward
        elif is_valid(x, y + 1, N) and not explored[x][y + 1]:
            y = y + 1  # Move right
        elif is_valid(x - 1, y, N) and not explored[x - 1][y]:
            x = x - 1  # Move backward
        elif is_valid(x, y - 1, N) and not explored[x][y - 1]:
            y = y - 1  # Move left
        else:
            # If all neighboring rooms are explored, move backward
            x = x - 1

        # Update charge and time taken
        charge -= 1
        time_taken += 1

    # Output total time taken (number of steps/moves- 1 step- 1 unit time taken)
    # Here total time I assumed to be the no. of steps or moves the agent makes 
    #while navigating the grid to find the pot of gold or until it decides to go to the charging station.
    print(f"Total time taken: {time_taken} units")

    # Output the location where the decision to charge was taken
    print(f"Decision to charge taken at: ({x}, {y})")

if __name__ == "__main__":
    random.seed()  

    # Inputs
    N = int(input("GRID SIZE (N): "))
    X = int(input("CAPACITY OF BATTERY (X): "))

    # Charging station location input
    charging_station = tuple(map(int, input("CHARGING STATION LOCATION (x y): ").split()))

    # Pot of gold location input
    pot_of_gold = tuple(map(int, input("GOLD POT LOACTION (x y): ").split()))

    # Agent Program
    agent_program(N, X, charging_station, pot_of_gold)
